samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -714,81 +744,92 @@ private void startSideInputs() {
-      taskInstanceMetrics.get(sideInputStorageManagers.get(ssp).getTaskName()).addOffsetGauge(
-          ssp, ScalaJavaUtil.toScalaFunction(() -> sideInputStorageManagers.get(ssp).getLastProcessedOffset(ssp)));
+    Map<TaskName, RunLoopTask> taskMap = new HashMap<>();
+    this.taskSideInputStoreSSPs.forEach((taskName, storesToSSPs) -> {
+        if (storesToSSPs.isEmpty()) {
+          return;
+        }
+        Set<SystemStreamPartition> taskSSPs = taskSideInputStoreSSPs.get(taskName).values().stream()
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
+
+        Map<String, SideInputsProcessor> sideInputsProcessors = taskSideInputStoreProcessors.get(taskName);
+
+        TaskSideInputHandler taskSideInputHandler = new TaskSideInputHandler(
+            taskName,
+            this.taskSideInputStorageManagers.get(taskName),
+            this.taskSideInputStoreSSPs.get(taskName),
+            sideInputsProcessors,
+            this.systemAdmins,
+            this.streamMetadataCache,
+            this.sideInputTasksCaughtUp,
+            this.initialSideInputSSPMetadata,
+            this.sideInputSSPLocks,
+            this.sideInputSSPCheckpointOffsets);
+        this.taskSideInputHandlers.put(taskName, taskSideInputHandler);
+
+        taskSSPs.forEach(ssp -> {
+            String startingOffset = taskSideInputHandler.getStartingOffset(ssp);
+
+            if (startingOffset == null) {
+              throw new SamzaException("No offset defined for SideInput SystemStreamPartition : " + ssp);
+            }
 
-      SystemStreamMetadata systemStreamMetadata = streamMetadataCache.getSystemStreamMetadata(ssp.getSystemStream(), false);
-      SystemStreamMetadata.SystemStreamPartitionMetadata sspMetadata =
-          (systemStreamMetadata == null) ? null : systemStreamMetadata.getSystemStreamPartitionMetadata().get(ssp.getPartition());
+            // register startingOffset with the sysConsumer and register a metric for it
+            sideInputSystemConsumers.register(ssp, startingOffset);
+            taskInstanceMetrics.get(taskName).addOffsetGauge(
+                ssp, ScalaJavaUtil.toScalaFunction(() -> taskSideInputHandler.getLastProcessedOffset(ssp)));
+          });
 
-      // record a copy of the sspMetadata, to later check if its caught up
-      initialSideInputSSPMetadata.put(ssp, sspMetadata);
+        taskMap.put(taskName, new SideInputRestoreTask(taskName, taskSSPs, taskSideInputHandler, taskSideInputStorageManagers.get(taskName), this.taskInstanceMetrics.get(taskName)));
+      });
 
-      // check if the ssp is caught to upcoming, even at start
-      checkSideInputCaughtUp(ssp, startingOffset, SystemStreamMetadata.OffsetType.UPCOMING, false);
-    }
+    TaskConfig taskConfig = new TaskConfig(config);
+    sideInputRunLoop = new RunLoop(taskMap,
+        null,
+        sideInputSystemConsumers,
+        1, // restrict task concurrency to 1 for now
```

Comment:
minor: suggest adding comments to other parameters in the construction that take raw values just like this.